### PR TITLE
Fix intrinsic lookup with namespaces

### LINKED ISSUE #7495: [Shader Execution Reordering] using namespace dx; not working for MaybeReorderThread
**Description**
using namespace dx; then calling MaybeReorderThread() produces different DXIL than dx::MaybeReorderThread.  The namespace version fails state object creation "Unresolved reference to function "MaybeReorderThread" (mangled name: "\01?MaybeReorderThread@dx@@YAXII@Z") by export "MyRaygenShader". "

**Steps to Reproduce**
Compile this distilled repro shader: dxc /t lib_6_9 NamespaceRepro.hltl.txt
[NamespaceRepro.hlsl.txt](https://github.com/user-attachments/files/20496303/NamespaceRepro.hlsl.txt)

Compare with compiling the same shader with "#define REPRO" commented out.

**Actual Behavior**
State object creation fails per above.

**Environment**
- DXC version 1.9 - 1.8.2505.28 (9efbb6c32); dxil.dll: 1.9(1.8.2505.28)
- Host Operating System Win11

### PR DESCRIPTION
**Pull request by llvm-beanz**
This change fixes issues with intrinsic lookup caused by not correctly respecting the using declaration(s) that impact unqualified lookups.

This probably isn't a perfect solution because I'm sure there's some nuance of unqualified lookups in C++ that I'm not handling, but this does respect scoped using directives and allows us to get things working.

Additionally this change disables emitting some "declared here" notes when the source location referred to is invalid.

Fixes #7495

### CODE REVIEW

File: tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        HitObject::TraceRay(Scene, RAY_FLAG_NONE, ~0, 0, 1, 0,
+                            ray, payload);
+
+    int sortKey = 1;
+    MaybeReorderThread(sortKey, 1);
+  }
+
+  {
+    int sortKey = 1;
+    MaybeReorderThread(sortKey, 1); // expected-error{{use of undeclared identifier 'MaybeReorderThread'; did you mean 'MaybeReorderThread'?}}
```
Comment (tex3d) [inline tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl:31]: Should the `did you mean 'MaybeReorderThread'` part say `did you mean 'dx::MaybeReorderThread'` instead?  Is there any way to get the message to say that?

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 2]
```diff
@@ -5204,81 +5360,121 @@ class HLSLExternalSource : public ExternalSemaSource {
...
+      SearchTables.push_back(std::make_pair(GlobalIntrinsics, m_hlslNSDecl));
+    else if (!isQualified) {
+      // If the name isn't qualified, we need to search all scopes that are
+      // accessible without qualification. This starts with the global scope and
+      // extends into any scopes that are referred to by using declarations.
+      SearchTables.push_back(std::make_pair(GlobalIntrinsics, m_hlslNSDecl));
+
+      // If we have a scope chain, walk it to get using declarations.
+      if (S) {
+        UnqualUsingDirectiveSet UDirs;
```
Comment (tex3d) [inline tools/clang/lib/Sema/SemaHLSL.cpp]: It seems unfortunate to have to look for all using directives for each call lookup.  Is there any way to make that better?  Shouldn't there be a better way to recognize the namespaces available in the current lookup scope based on using statements?

Comment (tex3d) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Ok, so I guess that it copies the `UnqualUsingDirectiveSet` implementation from `SemaLookup.cpp`.

I see we are duplicating this because the ExternalSemaSource mechanism seems to only support unqualified name lookups, as opposed to being hooked in under CppNamespaceLookup/LookupDirect (I think).

Do you think it would be a better approach to hook into the existing C++ lookup process differently?

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 3]
```diff
@@ -5204,81 +5360,121 @@ class HLSLExternalSource : public ExternalSemaSource {
...
+        // Find the first namespace or translation-unit scope.
+        Scope *Initial = S;
+        while (S && !isNamespaceOrTranslationUnitScope(S))
+          S = S->getParent();
```
Comment (tex3d) [inline tools/clang/lib/Sema/SemaHLSL.cpp]: I would think we need to visit each (non-transparent) scope on the way up to add the available namespaces, like:
https://github.com/microsoft/DirectXShaderCompiler/blob/94abfe972ad839185965f670329bcf33cd7bccbd/tools/clang/lib/Sema/SemaLookup.cpp#L1054-L1059

Comment (tex3d) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Ok, I see now that visitScopeChain should add these, and this just gets the `InnermostFileScope` for that call.

I think it would have been clearer to declare a new `Scope *InnermostFileScope = S;`, which would be changed by this loop, instead of changing the original passed-in `S`.

Comment (tex3d) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: I'm also a little confused.  The code I pointed to in `SemaLookup.cpp` is followed by the equivalent code to this.  So why does the code in `SemaLookup.cpp` explicitly visit each context up the parent chain before finding `InnermostFileScope` calling `visitScopeChain`?

It appears necessary, since `InnermostFileScope` will stop at the first global/namespace parent scope, while the prior loop will catch all scopes that could have using statements that impact the current lookup.

File: tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl (hunk 1 above)
Comment (llvm-beanz) [reply tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl:31]: There is something wrong with the way we're generating the ASTs for these types, and I'm not entirely sure how to fix it. I can carve out some time to look at it, but if it proves to be too time consuming I'd like to suggest tracking that as a follow up?

Does that sound reasonable to you?

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 2 above)
Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Switching to use the C++ lookup process would require us to pre-populate all the impacted APIs in the AST and now allow them to be lazily added (at least not without additional complicated changes). That might not have too significant of a performance impact because we don't have many intrinsics that are in namespaces today.

One drawbacks here is that the candidate selection and overload resolution are pretty tightly coupled in the codebase, so we'd really be opting in a subset of functions to the C++ overload resolution behavior. We do this today for subscript operators, which is fine because HLSL and C++ overload resolution for functions that take a single argument are almost indistinguishable in most cases. Extending this across a wider array of functions could be more problematic in the long run. It is _probably_ fine for the current set of functions, but would be an odd edge case we could trip over.

I think these concerns make me think it is probably better to teach the HLSL-specific code to handle unqualified lookups. WDYT @tex3d?

File: tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl (hunk 1 above)
Comment (tex3d) [reply tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl:31]: Yeah, that sounds reasonable.

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 2 above)
Comment (tex3d) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: So I was thinking of some way of hooking into the same code path that already collects the needed `using` namespaces as for other lookups, and exposing that namespace set to our lookup somehow.  Like by moving our lookup to `CppNamespaceLookup` instead of under `LookupName`.  It seems like the ExternalSource mechanism was only designed to deal with unqualified C names, as opposed to things that reside in C++ namespaces.

I wasn't thinking of adding the prototypes in advance so that the normal lookups find those.

However, that does bring up another issue.  We currently pre-populate `dx` and `vk` namespaces using `AddIntrinsicFunctionsToNamespace`, then we go through the usual path to construct a template/specialization in the usual way in `AddOverloadedCallCandidates` for those functions later.  This pre-population path was originally added for `vk::RawBuffer*` intrinsics, and has specific limitations for the template types it recognizes from the intrinsic table as a result (see `CreateTemplateTypeParmDeclsForIntrinsicFunction`).  Shouldn't this path be unnecessary with this change?

Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: I'm not sure I understand what you're saying here. `Sema::LookupName` calls `CppLookupName`, which calls `CppNamespaceLookup`. The external sema source gets called in all cases where `LookupName` fails, so it should always be invoked for both C and C++ names.

Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: So... looking at this more. Part of the problem is that we're using the ExternalSemaSource wrong.

DXC added a new interface to the ExternalSemaSource to populate overload candidates, but we should instead probably have handled it all in `lookupUnqualified`, which is really the right place to be able to perform a lazy lookup. We really only depend on `lookupUnqualified` for a few simple cases for builtin types, but it could do so much more.

I'm not really sure it is worth fixing this all in DXC.

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 3 above)
Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Oh... so that's interesting. This is needed for a C++-ism that I hadn't really thought about.

The `visitScopeChain` call hits the contexts up the scope chain, but it doesn't guarantee that it is walking up the declaration contexts for the enclosing scope. The commit that adds that code is [here](https://github.com/llvm/llvm-project/commit/cc9406c055e329dde1b1aeb610e449ae6981863a#diff-3daf0c578caa741d697a7fc479c7c7d4d27b4e502593d599ea6b513557de83ec), its test case demonstrates the issue:

```c++
namespace Other {
  void other_foo();
}

namespace M2 {
  using namespace Other;

  namespace MInner {
    class Bar { 
      void bar();
    };
  }
}

void M2::MInner::Bar::bar() {
  other_foo();
}
```

The issue is that the scope chain for `M2::MInner::Bar::bar()` doesn't actually include the `using namespace Other` declaration, you have to walk the declaration contexts for the enclosing decl.

Because namespaces aren't sealed, we could encounter this, so I'll add a declaration context walk to my PR as well.

File: tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl (hunk 1 above)
Comment (llvm-beanz) [reply tools/clang/test/SemaHLSL/using-namespace-dx-errors.hlsl:31]: I have a follow-up change ready that makes this less misleading by removing the correction text. That falls out naturally from fixing how SemaHLSL constructs ASTs to properly put namespaced functions under the namespace instead of the translation unit declaration.

If you're curious to see that change it is [here](https://github.com/llvm-beanz/DirectXShaderCompiler/commit/e949e23fc5f9a0b70388031b39a1e8f7ba783ddb).

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 4]
```diff
@@ -5204,81 +5207,121 @@ class HLSLExternalSource : public ExternalSemaSource {
...
+
+    llvm::SmallVector<IntrinsicTableEntry, 3> SearchTables;
+
+    if (isDxNamespace)
+      SearchTables.push_back(
+          IntrinsicTableEntry{IntrinsicArray(g_DxIntrinsics), m_dxNSDecl});
+#ifdef ENABLE_SPIRV_CODEGEN
+    else if (isVkNamespace)
+      SearchTables.push_back(
+          IntrinsicTableEntry{IntrinsicArray(g_VkIntrinsics), m_vkNSDecl});
```
Comment (simoll) [inline tools/clang/lib/Sema/SemaHLSL.cpp:5241]: Maybe deduplicate the pushing to SearchTables by fusing this with the code below in the `!isQualified` case?

Comment (simoll) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5241]: (this would also make sure you always get the same table ordering)

**Comment by simoll** [comment]
Just one nit. Otherwise LGTM.

**Comment by llvm-beanz** [comment]
> Just one nit. Otherwise LGTM.

Not seeing any other comments. Did something get stuck in GitHub's UI?

**Review by simoll — APPROVED** [review]

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 4 above)
Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5241]: If the name is qualified, we only get one search table, so the ordering should always be consistent (although ordering explicitly cannot impact correctness because if we find more than one valid name it is an error since name lookup can't be ambiguous).

I'll take a stab at merging the code, but I'm unsure if it will actually make it easier to follow.

Comment (llvm-beanz) [reply tools/clang/lib/Sema/SemaHLSL.cpp:5241]: Oh, yea, the cleaned up version is much nicer. Pushing momentarily.

**Review by V-FEXrt — APPROVED** [review]
